digitalstrom_mqtt/mqtt.go

```diff
@@ -194,6 +224,254 @@ func (dm *DigitalstromMqtt) getTopic(deviceType string, deviceId string, deviceN
 	return topic
 }
 
+// Returns MQTT topic to publish the Server status.
+func (dm *DigitalstromMqtt) getStatusTopic() string {
+	root_topic := strings.Split(dm.config.TopicFormat, "/")[0]
+	return root_topic + "/server/state"
+}
+
+// Publish the current binary status into the MQTT topic.
+func (dm *DigitalstromMqtt) publishDiscoveryMessages() {
+	for _, device := range dm.digitalstrom.GetAllDevices() {
+		messages, err := dm.deviceToHomeAssistantDiscoveryMessage(device)
+		if utils.CheckNoErrorAndPrint(err) {
+			for _, discovery_message := range messages {
+				dm.client.Publish(discovery_message.topic, 0, dm.config.Retain, discovery_message.message)
+			}
+		}
+	}
+	for _, circuit := range dm.digitalstrom.GetAllCircuits() {
+		messages, err := dm.circuitToHomeAssistantDiscoveryMessage(circuit)
+		if utils.CheckNoErrorAndPrint(err) {
+			for _, discovery_message := range messages {
+				dm.client.Publish(discovery_message.topic, 0, dm.config.Retain, discovery_message.message)
+			}
+		}
+	}
+}
+
+// Define a Home Assistant discovery message as its MQTT topic and the message
+// to be published.
+type HassDiscoveryMessage struct {
+	topic   string
+	message []byte
+}
+
+// Return the definition of the light and cover entities coming from a device.
+func (dm *DigitalstromMqtt) deviceToHomeAssistantDiscoveryMessage(device digitalstrom.Device) ([]HassDiscoveryMessage, error) {
+	// Check for device instances where the discovery message can not be created.
+	if device.Name == "" {
+		return nil, fmt.Errorf("empty device name, skipping discovery message")
+	}
+	if (device.DeviceType != digitalstrom.Light) && (device.DeviceType != digitalstrom.Blind) {
+		return nil, fmt.Errorf("device type not supported %s", device.DeviceType)
+	}
+	device_config := map[string]interface{}{
+		"configuration_url": "https://" + dm.ds_config.Host,
+		"identifiers":       []interface{}{device.Dsid, device.Dsuid},
+		"manufacturer":      "DigitalStrom",
+		"model":             device.HwInfo,
+		"name":              device.Name,
+	}
+	availability := []interface{}{
+		map[string]interface{}{
+			"topic":                 dm.getStatusTopic(),
+			"payload_available":     Online,
+			"payload_not_available": Offline,
+		},
+	}
+	message := map[string]interface{}{}
+	topic := ""
+	if device.DeviceType == digitalstrom.Light {
+		// Setup configuration for a MQTT Cover in Home Assistant:
+		// https://www.home-assistant.io/integrations/light.mqtt/
+		topic = dm.config.HomeAssistantDiscoveryPrefix + "/light/" + device.Dsid + "/light/config"
+		message = map[string]interface{}{
+			"device":            device_config,
+			"name":              utils.RemoveRegexp(device.Name, "light"),
+			"unique_id":         device.Dsid + "_light",
```

Comment:
isn't the id enough ? why do you have to add a postfix ?

Reply:
This postfix is because a device and an entity in Home assistant are two different things. 1 device could have multiple entities. In this case the `DSID` is the physical device ID, and therefore we need to make sure that all entities have uniqueness and for that reason I add the suffix to generate the `unique_id`. I know that the lights devices right now expose a single entity, but this is taking into account that in the future, we may want to export more entities from that device (e.g. sensor about the signal intensity of the device, the push button from the GE device,...). Though now it may not make any sense, this is thinking about the future which changing the `unique_id` if needed could mess up with your Home Assistant setup.

Reply:
I see your point but I don't see digitalstrom drastically changing anything soon.

Again, in my opinion, I really don't think it's necessary. But if you think it could then fine.

Reply:
But in this case better be safe than sorry. In case more entities can be exposed per device, then there would be the breaking change of having to update the unique ID of the entities. This is the approach that takes some other projects that expose Home Automation devices through MQTT like zigbee2mqtt.